cli/src/runner.rs

```diff
@@ -125,29 +126,49 @@ pub async fn extract_failed_tests(
                 }
             };
             let reader = std::io::BufReader::new(file);
-            let junitxml = match junit_parser::from_reader(reader) {
+            let mut junitxml = JunitParser::new();
+            match junitxml.parse(reader) {
                 Ok(junitxml) => junitxml,
                 Err(e) => {
                     log::warn!("Error parsing junitxml: {}", e);
                     continue;
                 }
             };
-            for suite in junitxml.suites {
-                let parent_name = suite.name;
-                for case in suite.cases {
-                    let failure = case.status.is_failure();
-                    if failure {
-                        let name = case.original_name;
-                        let test = Test::new(
-                            name.clone(),
-                            parent_name.clone(),
-                            case.classname.clone(),
-                            case.file.clone(),
-                            org_slug,
-                            repo,
-                        );
-
-                        failures.push(test);
+            for report in junitxml.reports() {
+                let test_suites = report.test_suites.clone();
+                for suite in test_suites {
+                    let parent_name = suite.name.into_string();
+                    for case in suite.test_cases {
+                        let is_failure = match case.status {
+                            TestCaseStatus::NonSuccess { .. } => true,
+                            _ => false,
+                        };
+                        if is_failure {
+                            let name = case.name.into_string();
+                            let classname = match case.classname {
+                                Some(classname) => Some(classname.into_string()),
+                                None => None,
+                            };
+                            let file = match case.extra.get("file") {
+                                Some(file) => Some(file.clone().into_string()),
+                                None => None,
+                            };
+                            let id = match case.extra.get("id") {
+                                Some(id) => Some(id.clone().into_string()),
+                                None => None,
+                            };
+                            let test = Test::new(
+                                name,
+                                parent_name.clone(),
+                                classname,
+                                file,
+                                id,
+                                org_slug,
+                                repo,
+                            );
+
+                            failures.push(test);
+                        }
                     }
                 }
             }
```

Comment:
There's a little more copying than necessary here. We can iterate by reference and clone only what's needed. Also, I sprinkled in some small changes to make control flow less verbose.

```suggestion
            for report in junitxml.reports() {
                for suite in &report.test_suites {
                    let parent_name = String::from(suite.name.as_str());
                    for case in &suite.test_cases {
                        if !matches!(case.status, TestCaseStatus::NonSuccess { .. }) {
                            continue;
                        }

                        let name = String::from(case.name.as_str());
                        let classname = case.classname.as_ref().map(|s| String::from(s.as_str()));
                        let file = case.extra.get("file").map(|s| String::from(s.as_str()));
                        let id = case.extra.get("id").map(|s| String::from(s.as_str()));
                        let test = Test::new(
                            name,
                            parent_name.clone(),
                            classname,
                            file,
                            id,
                            org_slug,
                            repo,
                        );

                        failures.push(test);
                    }
                }
            }
```

Also, if wanted, we can be fancy and do something like this too:
```rs
let xml_string_to_string = |s: &quick_junit::XmlString| String::from(s.as_str());
let classname = case.classname.as_ref().map(xml_string_to_string);
let file = case.extra.get("file").map(xml_string_to_string);
let id = case.extra.get("id").map(xml_string_to_string);
```

Reply:
I like your style.